src/Simulation/qdk_sim_rs/src/c_api.rs

```diff
@@ -0,0 +1,443 @@
+pub unsafe extern "C" fn m(sim_id: usize, idx: usize, result_out: *mut usize) -> i64 {
+    as_capi_err(|| {
+        let state = &mut *STATE.lock().unwrap();
+        if let Some(sim_state) = state.get_mut(&sim_id) {
+            let instrument = &sim_state.noise_model.z_meas;
+            let (result, new_state) = instrument.sample(&[idx], &sim_state.register_state);
+            sim_state.register_state = new_state;
+            *result_out = result;
+            Ok(())
+        } else {
+            Err(format!("No simulator with id {} exists.", sim_id))
+        }
+    })
+}
+
+/// Gets the noise model corresponding to a particular name, serialized as a
+/// string representing a JSON object.
+///
+/// Currently recognized names:
+/// - `ideal`
+/// - `ideal_stabilizer`
+///
+/// # Safety
+/// As this is a C-language API, the Rust compiler cannot guarantee safety when
+/// calling into this function. The caller is responsible for ensuring that:
+///
+/// - `name` is a pointer to a null-terminated string of Unicode characters,
+///   encoded as UTF-8, and that the pointer remains valid for the lifetime of
+///   this call.
+/// - `noise_model_json` is a valid pointer whose lifetime extends for the
+///    duration of this function call.
+///
+/// After this call completes, this function guarantees that either of the two
+/// conditions below holds:
+///
+/// - The return value is negative, in which case calling `lasterr` will return
+///   an actionable error message, or
+/// - The return value is `0`, and `*noise_model_json` is a valid pointer to a
+///   null-terminated string of Unicode characters, encoded as UTF-8. In this
+///   case, the caller is considered to own the memory allocated for this
+///   string.
+#[no_mangle]
+pub extern "C" fn get_noise_model_by_name(
+    name: *const c_char,
+    noise_model_json: *mut *const c_char,
+) -> i64 {
+    as_capi_err(|| {
+        let name = unsafe { CStr::from_ptr(name) }
+            .to_str()
+            .map_err(|e| format!("UTF-8 error decoding representation argument: {}", e))?;
+        let noise_model = NoiseModel::get_by_name(name)?;
+        let noise_model = CString::new(noise_model.as_json()).unwrap();
+        unsafe {
+            *noise_model_json = noise_model.into_raw();
+        }
+        Ok(())
+    })
+}
+
+/// Returns the currently configured noise model for a given simulator,
+/// serialized as a string representing a JSON object.
+///
+/// # Safety
+/// As this is a C-language API, the Rust compiler cannot guarantee safety when
+/// calling into this function. The caller is responsible for ensuring that:
+///
+/// - `noise_model_json` is a valid pointer whose lifetime extends for the
+///    duration of this function call.
+///
+/// After this call completes, this function guarantees that either of the two
+/// conditions below holds:
+///
+/// - The return value is negative, in which case calling `lasterr` will return
+///   an actionable error message, or
```

Comment:
Nit: Returning an arbitrary negative value as an indication of an error, limits the space for the subsequent maneuver. 

At some point in the future we may want to distinguish one group of failures from the other group, by returning `-1` vs. `-2`. You may want to consider limiting the error indication to one value (e.g. `-1`).
(to keep in mind for the future)

Reply:
Fair; in this case, all failures get translated to the same C# exception, but I agree, we can consider providing more granular failure codes in the future.

Reply:
https://github.com/microsoft/qsharp-runtime/issues/732